src/azure-cli/azure/cli/command_modules/role/custom.py

```diff
@@ -854,37 +836,77 @@ def list_permission_grants(cmd, identifier=None, query_filter=None, show_resourc
     return result
 
 
-def add_permission(cmd, identifier, api, api_permissions):
+def add_required_permission(cmd, identifier, api, api_permissions):
+    """Adds a required permissions for an API to requiredResourceAccess"""
+
     graph_client = _graph_client_factory(cmd.cli_ctx)
+    api_app_id = _resolve_api_identifier_to_app_id(graph_client, api)
     application = show_application(graph_client.applications, identifier)
-    existing = application.required_resource_access
-    resource_accesses = []
+    rras = application.required_resource_access
+
+    # If a required resource access for the API exists, use it. Otherwise, create a new one.
+    rra = next((e for e in rras if e.resource_app_id == api_app_id), None)
+    if rra is None:
+        rra = RequiredResourceAccess(resource_app_id=api_app_id, resource_access=[])
+        rras.append(rra)
+
+    # Add each required permission unless it already exists in the RRA.
     for e in api_permissions:
-        access_id, access_type = e.split('=')
-        resource_accesses.append(ResourceAccess(id=access_id, type=access_type))
+        permission_id, permission_type = e.split('=')
+        if not any((ra.id == permission_id and ra.type == permission_type) for ra in rra.resource_access):
```

Comment:
This is great change! If we don't detect duplication, we end up with duplicated entries in the request body:

```
> az ad app permission add --id 46eb4122-bd2b-4f54-af7b-6d79b46ee31a  --api 00000002-0000-0000-c000-000000000000 --api-permissions 311a71cc-e848-46a1-bdf8-97ff7156d8e6=Scope --debug

{
    "requiredResourceAccess": [
        {
            "resourceAccess": [
                {
                    "id": "311a71cc-e848-46a1-bdf8-97ff7156d8e6",
                    "type": "Scope"
                },
                {
                    "id": "311a71cc-e848-46a1-bdf8-97ff7156d8e6",
                    "type": "Scope"
                },
                {
                    "id": "311a71cc-e848-46a1-bdf8-97ff7156d8e6",
                    "type": "Scope"
                },
                {
                    "id": "311a71cc-e848-46a1-bdf8-97ff7156d8e6",
                    "type": "Scope"
                }
            ],
            "resourceAppId": "00000002-0000-0000-c000-000000000000"
        },
        {
            "resourceAccess": [
                {
                    "id": "c79f8feb-a9db-4090-85f9-90d820caa0eb",
                    "type": "Scope"
                },
                {
                    "id": "9a5d68dd-52b0-4cc2-bd40-abcf44ac3a30",
                    "type": "Role"
                }
            ],
            "resourceAppId": "00000003-0000-0000-c000-000000000000"
        }
    ]
}
```

However, the server should also be responsible for detecting duplication, right? 😉

Reply:
Bug opened internally to detect and avoid server-side duplicates. (It's not harmful in itself, but we can certainly reduce the chances of surprising a client which doesn't handle this edge case correctly.)